### ISSUE TITLE
Pin valhalla dependencies

### ISSUE THREAD

**Issue by mandeepsandhu** [valhalla/valhalla#2124]
Recently, Valhalla build was broken when the version of `geos` was upgraded to `3.8.0`. Ideally, we want full control over what version of the dependencies get installed. This ensures we don't break functionality when external project makes incompatible changes.

**Comment by kevinkreiser** [comment]
Pinning versions of all the dependencies also makes it very very hard to support a wide range of build environments. Inside of `CMakeLists.txt` (and in configure.ac before that) we often set requirements using something like `>=` for APIs that are pretty stable. This has the benefit that most people can just use whatever package manager they have to get whatever packages are needed and so long as they arent running centos4 they'll be able to meet the version requirements. For other things we will simply vendor the sources (ie. if the license allows it) to avoid any issues.

Specifically for data building (ie where we have extra dependencies like libgeos, libspatialite, sqlite, lua, etc) we have been pretty lax on dependency version control because we dont envision people wanting to build tile data on arm cpus for example. In other words we expect you to pretty much use ubuntu to build tiles. At any rate, sadly the c++ port of libgeos, and also the c++ apis for protobuf have given us trouble in these regards but again we are trying to keep this code accessable for as many people as possible. Since we cannot foresee when an api like libgeos or protobuf will make breaking changes we have taken the stance that when they do, builds will fail and we'll have to deal with it. Either by telling the person to get a previous version or by doing some preprocessor fun to pick the right implementation.

Just to be clear our builds were not broken by libgeos 3.8.0 as none of our build systems were using that version of the dependency. In other words, we have full controll over our build system so we wont be arbitrarily updating our dependencies without knowing it. What happened was, some users with macs using homebrew were stuck because they had had 3.8.0. They complained, and @PureW did the preprocessor work.

**Comment by mandeepsandhu** [comment]
You're right, for a library, it makes sense not to pin dependencies (and only specify a minimum version that it wants). For applications, I think its best to lock down the deps to get consistent build artifacts. So if the main artifact of the build is a library, then we need not pin versions and rather leave it to the user of the library to decide what exact version of dependency they want.

I'm not sure what you meant by pinning versions restricting build environments. Do you mean if we pin a version of a dependency, some build environment (say yum/rpm based) might not have that specific version? If so, then wouldn't we want to gracefully fail the build? I might be completely missing your point here though! :)
 
Most decent libraries/project strive hard not to make breaking changes for their users. So you're right we shouldn't be planning on dependencies breaking all the time (and if they frequently do, we should look at stable alternatives).

Looking at @PureW 's change, it seems that it was done to handle a breaking change in geo's API. I'm not sure how this is a `brew` issue though. This could happen on a Linux env too if someone ran `apt upgrade` and  `libgeos` version got bumped to `3.8.x` (it would break build if @PureW 's change wasn't there).
We're also not installing `geos` explicitly (at least if you see how the mac deps are installed), and rather getting it as a dependency of `libsaptialite`. Maybe we should list out the `geos` dependency explicitly since we also use it directly in valhalla code?

We can close this issue if its not a concern (although right not clear to me as to why).

Thanks for your input!

**Comment by nilsnolde** [comment]
> This could happen on a Linux env too if someone ran apt upgrade and libgeos version got bumped to 3.8.x (it would break build if @PureW 's change wasn't there).

I can verify that (unfortunately I had the ubuntugis/unstable repo active): https://github.com/valhalla/valhalla/issues/2021#issuecomment-562479630

**Comment by PureW** [comment]
To me, there are two conflicting goals.

1) In order to build and run on as many platforms and systems as possible in good open-source fashion, it's useful to allow for a range of (compatible) dependencies.

2) As someone running a production service using a package built in this way, it is important to take pre-cautions so as to not introduce unwanted changes in the final build artifact due to changes upstream in dependencies.

A possible solution that allows for both of these goals to co-exist is to not make any changes to the valhalla project, but instead make sure the environment that builds the binaries is fully specified, perhaps using something like the nix package manager.

**Comment by mandeepsandhu** [comment]
I think some of the confusion/concern in this issue is due to the way I worded this issue, and I'm sorry I didn't do a good job of explaining what I meant.
I'm not advocating for pinning versions of the dependencies. Although that might be desirable for a specific dependency like `geos` because of our use of its unstable API.

I think all of us agree here that we only need to specify the minimum version of dependency we're known/tested to work with. Only in special cases where specifying just the min ver wont work, will we resort to pinning the dependency version.

@PureW , you're right, this task should be the responsibility of the build/packaging system that makes the valhalla lib. `nix` looks like a good candidate. I also stumbled upon [buckaroo](https://buckaroo.pm/) while googling around for package management systems. We should probably do a quick evaluation of a few candidates and settle on one.

**Comment by nilsnolde** [comment]
was just looking for geos tickets to close and found this one. to me it looks like we can close this right? mins versions of linked libs are set via cmake and since we don't publish many artifacts other than docker, IMO it's the responsibility of anyone building valhalla for production settings to make sure their env is predictable. we only resorted to using conan as package manager for boost because it packages the headers in a friendly way and we wanted to somehow vendor it.
